### ISSUE TITLE
`defineFun` allows neither to reference leading implicit variables nor bind them by hand.

### ISSUE THREAD

**Issue by effectfully** [agda/agda#2191]
Consider this function

```
typeOf : {A : Set} -> A -> Set
typeOf {A} _ = A
```

We may try to define it via reflection as follows:

```
open import Data.List.Base
open import Reflection

fun = lam visible (abs "_" (var 1 []))

typeOfViaDefineFun : Name -> TC _
typeOfViaDefineFun f = defineFun f (clause [] fun ∷ [])

failTypeOfViaDefineFun : {A : Set} -> A -> Set
unquoteDef failTypeOfViaDefineFun = typeOfViaDefineFun testTypeOfViaDefineFun
```

Agda throws `de Bruijn index 1 is not in scope in the context (z : _)`. So `defineFun` is not aware of leading implicit variables. OK, let's redefine `fun` so the implicit variable is introduced explicitly:

`fun = lam hidden (abs "_" (lam visible (abs "_" (var 1 []))))`

An attempt to type check `failTypeOfViaDefineFun` with this definition of `fun` in scope results in

```
Found an implicit lambda where an explicit lambda was expected
when checking that the expression λ {z} → λ z₁ → z has type
.A → Set
```

I'd prefer `defineFun` to be aware of leading implicit arguments (as opposed to not truncating type signatures).

**Comment by UlfNorell** [comment]
This isn't a reflection problem. Your first attempt doesn't really make sense: there really is no binding for `var 1`. The second attempt doesn't work if you try it without reflection either:

``` agda
f : {A : Set} → Set → Set
f = λ {A} _ → A -- Found implicit lambda where explicit was expected
```

If you introduce the implicit in the left-hand side there are no problems.

**Comment by effectfully** [comment]
> This isn't a reflection problem. Your first attempt doesn't really make sense: there really is no binding for `var 1`.

```
fun = lam visible (abs "_" (var 1 []))

macro
  typeOfViaMacro : Term -> TC _
  typeOfViaMacro = unify fun

testTypeOfViaMacro : {A : Set} -> A -> Set
testTypeOfViaMacro = typeOfViaMacro
```

Type checks perfectly. `var 1` refers to silently binded `{A}` and I don't see why would `defineFun` behave not like everything else. Here is another example:

```
fun = lam visible (abs "_" (var 1 []))

typeOf′ : {A : Set} -> A -> Set
typeOf′ = unquote (unify fun)
```

OK, you may want `defineFun` to be able to define only closed terms. Why doesn't my second attempt type check then? `{A : Set} → A → Set` gets truncated to `A → Set` — Agda pretends she silently introduced an implicit variable, but she doesn't allow to reference it. I insist it's a reflection bug.

**Comment by UlfNorell** [comment]
Oh, I see what you mean. I'm not sure it's a good idea to fix though. When you say

``` agda
foo : {A : Set} → A → Set
foo = whatever
```

the `A` is introduced in the left-hand side. Thus it's equivalent to

``` agda
foo : {A : Set} → A → Set
foo {A} = whatever
```

If `whatever` is a macro it can refer to the silently bound `A` by its index.

The problem with allowing this also for `defineFun` is that it becomes quite hard to predict what deBruijn index to use for a given variable, since you have to count silently introduced variables in the left-hand side. And if you are able to count the missing variables, it's just as easy to make them explicit.

**Comment by effectfully** [comment]
> The problem with allowing this also for defineFun is that it becomes quite hard to predict what deBruijn index to use for a given variable, since you have to count silently introduced variables in the left-hand side.

I'm not sure I'm following. `defineFun` requires a function to be already declared, so its type is known and you don't need to make any predictions. Referencing leading implicits doesn't affect your definition, because all other variables are introduced later and have lower indices therefore.

I noticed this behaviour when I tried to refactor a reflected macro application into `defineFun` and got "de Bruijn index is not in scope". If you don't want to introduce implicits silently, then it should be consistent: you can't reference them at the value level, but they also shouldn't be truncated at the type level. But as far as I see allowing to reference leading implicits in `defineFun` is consistent with other parts of reflection and is also backwards compatible.

**Comment by UlfNorell** [comment]
> I'm not sure I'm following. defineFun requires a function to be already declared, so its type is known and you don't need to make any predictions. If you don't reference leading implicits, your definition is unaffected, because all other variables are introduced later and have lower indices therefore.

That's not true. There will also be implicitly introduced constructor arguments that you need to take into account.

For value level macros it would make little sense to disallow referencing implicitly introduced variables from the context, since they may well show up in the type of later arguments.

**Comment by effectfully** [comment]
> That's not true. There will also be implicitly introduced constructor arguments that you need to take into account.

Sorry, I don't understand. Could you please show an example?

> For value level macros it would make little sense to disallow referencing implicitly introduced variables from the context, since they may well show up in the type of later arguments.

I mean the value level part of `defineFun`. This error

```
Found an implicit lambda where an explicit lambda was expected
when checking that the expression λ {z} → λ z₁ → z has type
.A → Set
```

indicates that `{A : Set}` was truncated from the type signature of `failTypeOfViaDefineFun`, but what is the reason to truncate it if no implicit is bound?

**Comment by UlfNorell** [comment]
> Sorry, I don't understand. Could you please show an example?

``` agda
vmap : ∀ {A B : Set} {n} → (A → B) → Vec A n → Vec B n
vmap f []       = []
vmap f (x ∷ xs) = f x ∷ vmap f xs
```

Currently, translating this to a reflected definition to be `defineFun`'d is straightforward. The deBruijn indices available in the right-hand side would be `0 = xs, 1 = x, 2 = f`. With your proposed change the index of `f` would change to `3` since there is an implicit length argument to `_∷_`.

> This error indicates that {A : Set} was truncated from the type signature of `failTypeOfViaDefineFun`, but what is the reason to truncate it if no implicit is bound?

The implicit `A` is bound but the reflected clause cannot refer to it, for the reasons above.

**Comment by effectfully** [comment]
> Currently, translating this to a reflected definition to be `defineFun`'d is straightforward.

Thanks, I now see what you mean. However I don't propose to silently bind all implicits — only the leading ones.

**Comment by UlfNorell** [comment]
That feels a little ad hoc, and I'm not sure what problem it solves.

From the discussion so far it seems that the main problem is more one of unintuitive behaviour rather than any actual problems with the current implementation. I feel like making it possible to refer to only leading implicits would make things even less intuitive.

Also, you couldn't do it for pattern lambdas without shifting the indices of outer variables.

**Comment by effectfully** [comment]
> That feels a little ad hoc, and I'm not sure what problem it solves.

It'd allow to treat patternless definitions (such definitions can contain explicit lambdas on rhs instead of bound variables on lhs) in a simple way. Currently when you generate a patternless definition (which happens pretty often for me), you can't neither prepend an implicit lambda to it, nor reference the implicit variable it binds. So you need to explicitly bind on lhs all implicit variables you want to reference, while leading implicits are already treated in a somewhat ad hoc way and introduced silently and eagerly everywhere. It feels natural to me to treat patternless definitions not as functions with absent pattern matching, but as aliases for terms, so in their bodies hold the same rules as for simple terms. The example I gave:

```
fun = lam visible (abs "_" (var 1 []))

typeOf′ : {A : Set} -> A -> Set
typeOf′ = unquote (unify fun)
```

Why should `var 1` be out of scope in patternless `typeOf′` defined via `defineFun`? These two ways to define `typeOf′` look equal to me.

**Comment by UlfNorell** [comment]
> you need to explicitly bind on lhs all implicit variables you want to reference

I still don't see why this is a big problem.

> leading implicits are already treated in a somewhat ad hoc way and introduced silently and eagerly everywhere

Actually it's trailing implicit arguments that are introduced eagerly. This has nothing to do with reflection though. See #2001 and #679. What makes you say that it's ad hoc?

> It feels natural to me to treat patternless definitions not as functions with absent pattern matching, but as aliases for terms, so in their bodies hold the same rules as for simple terms.

I'm not sure exactly what you mean by this, but the fact is that when you write

``` agda
typeOf : {A : Set} → A → Set
typeOf = ?
```

an implicit variable is inserted in the left-hand side and the goal type in the right-hand side is `.A → Set`.

> Why should `var 1` be out of scope in patternless `typeOf′` defined via `defineFun`? These two ways to define `typeOf′` look equal to me.

This argument applies to implicits anywhere, which I thought we already agreed would be a bad idea to allow you to reference.

**Comment by effectfully** [comment]
> I still don't see why this is a big problem.

It's not. I defined `defineTerm : Name -> Term -> TC _`  that silently introduces leading implicits and assigns a term to a name regardless of whether variables in the term refer to leading implicits or not. It's a matter of a few lines. I just thought this specific case deserves to be handled out of the box. Perhaps I should stop making noise. I was wrong when I said this behaviour is a bug.

> Actually it's trailing implicit arguments that are introduced eagerly. 

Yeah, I tried to say that in the case of patternless definitions leading implicits are introduced eagerly, because they are also trailing implicits, which are introduced eagerly everywhere.

> What makes you say that it's ad hoc?

Trailing implicits are the only things (right?) that are silently introduced without asking for it. I see how this is useful, though.
